File: Remove_Node_in_Binary_Search_Tree.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param root: The root of the binary search tree.
     * @param value: Remove the node with given value.
     * @return: The root of the binary search tree after removal.
     */
    TreeNode* removeNode(TreeNode* root, int value) {
        // write your code here
        if (!root)
        {
            return nullptr;
        }
        if (root->val > value)
        {
            root->left = removeNode(root->left, value);
        }
        else if (root->val < value)
        {
            root->right = removeNode(root->right, value);
        }
        else
        {
            if (!root->left)
            {
                return root->right;
            }
            else if (!root->right)
            {
                return root->left;
            }
            TreeNode *cp = root;
            root = findAndDelMaxLeft(root->left);
            root->right = cp->right;
        }
        return root;
    }
    
    TreeNode *findAndDelMaxLeft(TreeNode *root)
    {
        TreeNode *parent;
        TreeNode *cp = root;
        while (root && root->right)
        {
            parent = root;
            root = root->right;
        }
        if (root != cp)
        {
            parent->right = root->left;
            root->left = cp;
        }
        return root;
    }
};

File: Remove Node in Binary Search Tree.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    """
    @param root: The root of the binary search tree.
    @param value: Remove the node with given value.
    @return: The root of the binary search tree after removal.
    """    
    def removeNode(self, root, value):
        # write your code here
        
        if not root:
            return
        else:
            if root.val > value:
                root.left = self.removeNode(root.left, value)
                return root
            elif root.val < value:
                root.right = self.removeNode(root.right, value)
                return root
            else:
                if not root.right:
                    return root.left
                # elif not root.left:
                #     return root.right
                else:
                    pre, cur = None, root.right
                    while cur and cur.left:
                        pre, cur = cur, cur.left
                    if pre:
                        pre.left = cur.right
                    else:
                        root.right = cur.right
                    cur.left, cur.right = root.left, root.right
                    return cur
            
